src/lib/core/coio_file.c

```diff
@@ -506,7 +506,12 @@ coio_tempdir(char *path, size_t path_len)
 {
 	INIT_COEIO_FILE(eio);
 
-	const char *tmpdir = getenv("TMPDIR");
+	bool exceeded;
+	const char *tmpdir = getenv_safe("TMPDIR", path_len, &exceeded);
```

Comment:
As I see, you replaced all `getenv()` calls, which are marked as already guarded (in the commit message) with `getenv_safe()`. I guess it is to follow the same safe pattern. However you don't do that for just one occurence in `src/`:

```c
src/find_path.c:77:			snprintf(buf, sizeof(buf) - 1, "%s", getenv("_"));
```

Is it intentional?

Reply:
Yep, I left it out because I can't rewrite it to `getenv_safe` without changing the behaviour. Well, I can, but that would look awful.

Here the env value, however large, is truncated to the buffer size. In order to implement that with `getenv_safe`, I would have to find the real env value length, by increasing sizes passed to `getenv_safe` until it returns something. And only then pass that value to be truncated in the buf.

Reply:
Yes, I could, but AFAICS with current approach too long `_` will be returned truncated.
And if I return NULL it won't be returned at all.

Reply:
The code calling `find_path` (in main.cc) has a fallback mechanism - it uses `argv[0]` if `find_path` fails - so it should probably not be an issue.

Reply:
> Yes, I could, but AFAICS with current approach too long `_` will be returned truncated.

Aren't we return `NULL` from `getenv_safe()` in the case (not a truncated string)?

Reply:
> The code calling find_path (in main.cc) has a fallback mechanism - it uses argv[0] if find_path fails - so it should probably not be an issue.

Still, argv[0] could be arbitrarily long, so I'd leave the current approach. In worst case (when an attacker can influence both env "_" and argv[0]) we will at least end up with a truncated value of known size.

Reply:
> Aren't we return NULL from getenv_safe() in the case (not a truncated string)?

Yes, we return NULL. I was talking of `find_path` as a whole. AFAIU @locker proposed to not bother with the fact that `getenv_safe` returns NULL, and pass it over to the caller. My objection was that it's better to return a truncated string (like it's done now with plain getenv in this place), because a caller falls back to `argv[0]`, which has arbitrary length.

Reply:
In my understanding, truncation is an unsafe approach here, especially, when we're talking about truncation before `realpath()`, disregarding of a source of the data. For example, an attacker can influence on `argv[0]` / `_` and pass `/usr/bin/../bin/../bin/<...>/../bin/malware/../tarantool`. Let's assume the patch is validated somewhere on the upper level and the validation is passed, because the path points to tarantool executable. But when it is truncated it may point to `/usr/bin/malware`. In some cases tarantool reexecutes its executable and actually it'll execute the malware.

This scenario is very theoretical and I wrote it just to illustrate that truncation may lead to a very unexpected result in this case.

Reply:
> executable. But when it is truncated it may point to /usr/bin/malware. In some cases tarantool reexecutes its executable and actually it'll execute the malware.

Good point. Anyway, now there's no input validation, so the attacker might simply put `/usr/bin/malware` into argv[0], and that's it.

My point is neither of the options (truncate env "_", fall back to argv[0]) is safe, so let's stick with the one which at least guarantees `tarantool_bin` will be no longer than `PATH_MAX`.

Reply:
And now it is truncated by `realpath()` or `snprintf()`. Okay.

I'm sceptic about truncating it, but I'm not a professional in this area and don't know how exactly construct a scenario and which scenarios should be taken seriously and which are negligible. Ideally we should consult on it with authors of the original issue (I'll leave it up to you) and possibly not in the scope of this issue.

I propose to focus a bit on this particular `getenv()` presence in the commit message and highlight that we consider it as safe in the sense you wrote above.

Reply:
Now `getenv_safe()` is called here, so my main question was resolved.

I would still vote to write something about `find_path()` result in the commit message except just 'this place is safe', because it is safe in the 'return NULL if too large' sense only when it falls back to the environment variable, but not in other cases, where it is 'return truncated value if too large'.

I'll resolve the discussion and leave it up to you.